src/Runtime/pdf3/internal/generator/browser_session.go

```diff
@@ -0,0 +1,540 @@
+		// When we return here, we never get to the cleanup state in the defer block of `generatePdf` below.
+		// Since we haven't actually used any of the user data the cleanup phase is completely safe to ignore.
+		// So let's manually swap into it cleanup state so that outer loop can swap out of this
+		previousState := w.state.Swap(uint32(SessionStateCleaningUp))
+		assert.AssertWithMessage(
+			previousState == uint32(SessionStateGenerating),
+			"Internal state should always go from Generating -> CleaningUp during early exit",
+		)
+		return
+	}
+
+	err := w.generatePdf(req)
+
+	if err != nil {
+		req.tryRespondError(mapCustomError(err), w)
+	}
+}
+
+func (w *browserSession) generatePdf(req *workerRequest) error {
+	request := req.request
+
+	// Ensure cleanup always runs
+	defer func() {
+		// When we get here we can already accept a request into the queue
+		// Cleanup should run fairly fast (low ms range)
+		previousState := w.state.Swap(uint32(SessionStateCleaningUp))
+		assert.AssertWithMessage(
+			previousState == uint32(SessionStateGenerating),
+			"Internal state should always go from Generating -> CleaningUp when starting cleanup",
+		)
+
+		// Navigate back to default
+		start := time.Now()
+		_, err := w.sendCommand("Page.navigate", map[string]interface{}{
+			"url": "about:blank",
+		})
+		if err != nil {
+			log.Printf("[%d, %s] failed to navigate out of url: %v\n", w.id, w.currentUrl, err)
+		}
+
+		if testInput := req.tryGetTestModeInput(); testInput != nil && testInput.CleanupDelaySeconds > 0 {
+			log.Printf("[%d, %s] waiting for %d seconds\n", w.id, w.currentUrl, testInput.CleanupDelaySeconds)
+			time.Sleep(testInput.CleanupDelaySeconds * time.Second)
+		}
+
+		// Cleanup browser storage
+		w.cleanupBrowser(req)
+
+		// Retry cleanup if it failed
+		if !req.cleanedUp {
+			log.Printf("[%d, %s] failed to cleanup storage, retrying...\n", w.id, w.currentUrl)
+			for range 3 {
+				w.cleanupBrowser(req)
+				if req.cleanedUp {
+					break
+				}
+			}
+
+			if !req.cleanedUp {
+				log.Fatalf("[%d, %s] failed to cleanup storage, we're in an unsafe state and can't proceed", w.id, w.currentUrl)
+			}
+		}
+		duration := time.Since(start)
+		log.Printf("[%d, %s] cleanup completed in %.2f seconds\n", w.id, w.currentUrl, duration.Seconds())
+	}()
+
+	// Set cookies
+	for _, cookie := range request.Cookies {
+		if req.ctx.Err() != nil {
+			return types.NewPDFError(types.ErrClientDropped, "", req.ctx.Err())
+		}
+
+		sameSite := "Lax"
+		switch cookie.SameSite {
+		case "Strict":
+			sameSite = "Strict"
+		case "None":
+			sameSite = "None"
+		}
+
+		_, err := w.sendCommand("Network.setCookie", map[string]interface{}{
+			"name":     cookie.Name,
+			"value":    cookie.Value,
+			"domain":   cookie.Domain,
+			"path":     "/",
+			"secure":   false,
+			"httpOnly": false,
+			"sameSite": sameSite,
+		})
+		if err != nil {
+			req.tryRespondError(types.NewPDFError(types.ErrSetCookieFail, "", err), w)
+			return nil
+		}
+	}
```

Comment:
If there are multiple cookies it is possible to set them all in a single cdp message instead of sending one message for each cookie: https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-setCookies